### PR TITLE
[MNOE-447] Dashboard templates management

### PR DESCRIPTION
**Pull request by ouranos**

### CODE REVIEW

**Comment by cesar-tonnoir** [comment]
@ouranos the PR is ready - can you please review?

File: api/app/controllers/mno_enterprise/jpi/v1/admin/impac/kpis_controller.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  # From the Admin panel, an admin can create/update/delete kpis to/from a dashboard template
+  class Jpi::V1::Admin::Impac::KpisController < Jpi::V1::Admin::BaseResourceController
+    before_filter :find_valid_kpi, only: [:update, :delete]
+
+    # POST /mnoe/jpi/v1/admin/impac/dashboard_templates/:id/kpis
+    def create
+      return render json: { errors: { message: 'Dashboard template not found' } }, status: :not_found unless template.present?
+
+      @kpi = template.kpis.create(kpi_create_params)
+      return render json: { errors: (kpi && kpi.errors).to_a }, status: :bad_request unless kpi.present? && kpi.valid?
```
Comment (ouranos) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/impac/kpis_controller.rb]: Should you use `@kpi` here? Just for consistency/readability?

File: api/app/controllers/mno_enterprise/jpi/v1/admin/impac/kpis_controller.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    def kpi_update_params
+      whitelist = [:element_watched, {extra_watchables: []}]
+      extract_params(whitelist)
+    end
+
+    def extract_params(whitelist)
+      (p = params).require(:kpi).permit(*whitelist).tap do |whitelisted|
+        whitelisted[:settings] = p[:kpi][:metadata] || {}
+        # TODO: strong params for targets & extra_params attributes (keys will depend on the kpi).
+        whitelisted[:targets] = p[:kpi][:targets] unless p[:kpi][:targets].blank?
```
Comment (ouranos) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/impac/kpis_controller.rb]: Could you do `if p[:kpi][:targets].present?`, affirmation are mentally easier to parse than negation

File: api/app/controllers/mno_enterprise/jpi/v1/admin/impac/widgets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    def widget
+      @widget ||= MnoEnterprise::Impac::Widget.find(params[:id].to_i)
+    end
+
+    def widget_create_params
+      params.require(:widget).permit(:endpoint, :name, :width).tap do |whitelisted|
+        whitelisted[:settings] = params[:widget][:metadata] || {}
+        # TODO: remove when mnohub migrated to new model
```
Comment (ouranos) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/impac/widgets_controller.rb]: Remove? Or add details?

File: api/app/controllers/mno_enterprise/jpi/v1/admin/impac/kpis_controller.rb (hunk 2 above)
Comment (cesar-tonnoir) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/impac/kpis_controller.rb]: :+1: look how I'm caring about your mental health :p

File: api/app/controllers/mno_enterprise/jpi/v1/admin/impac/widgets_controller.rb (hunk 3 above)
Comment (cesar-tonnoir) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/impac/widgets_controller.rb]: actually it's the other way around: we have to make sure all impac angular are >= 1.5.0

**Pull request merged**